lib/src/xml/grammar.dart

```diff
@@ -38,22 +38,25 @@ abstract class XmlGrammarDefinition<TNode, TName>
   // Connects the productions and the XML AST callbacks.
 
   @override
-  Parser attribute() => super
-      .attribute()
-      .map((each) => createAttribute(each[0], each[4][0], each[4][1]));
+  Parser attribute() => super.attribute().map((each) {
+        each as List<dynamic>;
```

Comment:
This PR is incomplete (~30 dynamic calls remain) but I wanted to get early feedback, @renggli : do these changes look acceptable to you? I can change the style of these fixes of course. I'd prefer to put `each[4]` into a typed variable, but I don't know what I'd call it. 😛

Reply:
The statement looks a bit lost :-)

Could the type be added as the function param? Or maybe the super method could be correctly typed instead?

Reply:
I don't know about the Parser class, like what type args it should have... but `super.attribute()` returns a `Parser<dynamic>`, and `map`'s signature is `Parser<R> map<R>(Callback<T, R> callback, {bool hasSideEffects = false})` where `T` is the type arg on Parser.

The callback (`(each) ...`) cannot be typed as `(List each) ...` because that doesn't match `map`'s signature.

The pre-null safe way to promote a value like this would be something like

```(each) {
  var eachAsList = each as List;
```

but with null safety, a simple `as` expression will promote the `each` variable, and there is no need to create a new variable or think of a new name for a new variable.

Reply:
The `as` expressions certainly looks fine.

We could also rewrite as 

     `super.attribute().cast<List<dynamic>>().map((each) { ...`

or

    `super.attribute().castList<dynamic>().map((each) { ...`

but they both come at quite a hefty cost.

Reply:
That said, the actual type of super-implementation is `Parser<List<dynamic>>` instead of the declared `Parser<dynamic>`, so that could also be changed?

Aside: Really seems like using subclassing to specify production actions is not such a good idea in a typed world.

Reply:
Well, ignore that last comment: If the super implementation of `attribute` would return a `List`, then the specialisation cannot return the AST node. Sorry, this is really messed up :-)

Reply:
> That said, the actual type of the super-implementation is `Parser<List<dynamic>>` instead of the declared Parser, so that could also be changed?

Since `class XmlEventDefinition extends XmlProductionDefinition`, this `attribute()` would also have to return `Parser<List<dynamic>>`, but it seems to return `Parser<XmlEventAttribute>`?

Reply:
OK Done.